webrender/res/cs_blur.glsl

```diff
@@ -5,131 +5,120 @@
+    gl_Position = uTransform * vec4(position, 0.0, 1.0);
 }
+
 #endif
 
 #ifdef WR_FRAGMENT_SHADER
 
-#if defined WR_FEATURE_COLOR_TARGET
+#define SUPPORT     4
+
+#ifdef WR_FEATURE_COLOR_TARGET
 #define SAMPLE_TYPE vec4
 #define SAMPLE_TEXTURE(uv)  texture(sCacheRGBA8, uv)
 #else
 #define SAMPLE_TYPE float
 #define SAMPLE_TEXTURE(uv)  texture(sCacheA8, uv).r
 #endif
 
-// TODO(gw): Write a fast path blur that handles smaller blur radii
-//           with a offset / weight uniform table and a constant
-//           loop iteration count!
-
-// TODO(gw): Make use of the bilinear sampling trick to reduce
-//           the number of texture fetches needed for a gaussian blur.
+// Accumulates two texels into the blurred fragment we're building up.
+void accumulate(float offset,
+                float crossAxisCoord,
+                inout vec2 gaussCoefficient,
+                inout SAMPLE_TYPE colorSum,
+                inout float factorSum) {
+    float factorA = gaussCoefficient.x;
+    gaussCoefficient *= vec2(gaussCoefficient.y, vCoefficients.y);
+    float factorB = gaussCoefficient.x;
+    gaussCoefficient *= vec2(gaussCoefficient.y, vCoefficients.y);
+
+    // Compute the texture coordinate that provides the correct linear combination of the two
+    // texels in question.
+    float factors = factorA + factorB;
+    float sampleOffset = offset + factorB / factors;
+
+    vec2 texCoord = vec2(sampleOffset, crossAxisCoord);
+    texCoord = clamp(vVertical != 0 ? texCoord.yx : texCoord.xy, vSrcRect.xy, vSrcRect.zw);
+
+    colorSum += factors * SAMPLE_TEXTURE(vec3(texCoord * vSrcSizeInv, vUv.z));
+    factorSum += factors;
+}
 
 void main(void) {
-    SAMPLE_TYPE original_color = SAMPLE_TEXTURE(vUv);
-
-    // TODO(gw): The gauss function gets NaNs when blur radius
-    //           is zero. In the future, detect this earlier
-    //           and skip the blur passes completely.
-    if (vBlurRadius == 0) {
-        oFragColor = vec4(original_color);
+    // FIXME(pcwalton): We shouldn't end up with zero blur radii in the first place!
+    if (vCoefficients.x == 0.0) {
+        vec2 texCoord = clamp(vUv.xy, vSrcRect.xy, vSrcRect.zw);
+        oFragColor = vec4(SAMPLE_TEXTURE(vec3(texCoord * vSrcSizeInv, vUv.z)));
         return;
     }
 
-    // Incremental Gaussian Coefficent Calculation (See GPU Gems 3 pp. 877 - 889)
-    vec3 gauss_coefficient;
-    gauss_coefficient.x = 1.0 / (sqrt(2.0 * 3.14159265) * vSigma);
-    gauss_coefficient.y = exp(-0.5 / (vSigma * vSigma));
-    gauss_coefficient.z = gauss_coefficient.y * gauss_coefficient.y;
-
-    float gauss_coefficient_sum = 0.0;
-    SAMPLE_TYPE avg_color = original_color * gauss_coefficient.x;
-    gauss_coefficient_sum += gauss_coefficient.x;
-    gauss_coefficient.xy *= gauss_coefficient.yz;
+    bool vertical = vVertical != 0;
+    vec2 axisCoord = vertical ? vUv.yx : vUv.xy;
+    float start = floor(axisCoord.x - float(SUPPORT)) + 0.5;
```

Comment:
if the flooring changes our start position, wouldn't the computed Gaussian factors no longer match?

Reply:
The computed Gaussian factors are based on `offset`, which is relative to the un-floored `axisCoord.x`.